src/Storages/PartitionStrategy.cpp

```diff
@@ -0,0 +1,262 @@
+
+    if (partition_strategy == "hive")
+    {
+        if (file_format.empty())
+        {
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "File format can't be empty for hive style partitioning");
+        }
+
+        return std::make_shared<HiveStylePartitionStrategy>(
+            partition_by,
+            sample_block,
+            context,
+            file_format,
+            hive_partition_strategy_write_partition_columns_into_files);
+    }
+
+    return std::make_shared<StringifiedPartitionStrategy>(partition_by, sample_block, context);
+}
+
+StringifiedPartitionStrategy::StringifiedPartitionStrategy(ASTPtr partition_by_, const Block & sample_block_, ContextPtr context_)
+    : PartitionStrategy(partition_by_, sample_block_, context_)
+{
+}
+
+StringifiedPartitionStrategy::PartitionExpressionActionsAndColumnName StringifiedPartitionStrategy::getExpression()
+{
+    PartitionExpressionActionsAndColumnName actions_with_column_name;
+
+    ASTs arguments(1, partition_by);
+    ASTPtr partition_by_string = makeASTFunction("toString", std::move(arguments));
+    auto syntax_result = TreeRewriter(context).analyze(partition_by_string, sample_block.getNamesAndTypesList());
+    actions_with_column_name.actions = ExpressionAnalyzer(partition_by_string, syntax_result, context).getActions(false);
+    actions_with_column_name.column_name = partition_by_string->getColumnName();
+
+    return actions_with_column_name;
+}
+
+std::string StringifiedPartitionStrategy::getPath(
+    const std::string & prefix,
+    const std::string & partition_key)
+{
+    return PartitionedSink::replaceWildcards(prefix, partition_key);
+}
+
+HiveStylePartitionStrategy::HiveStylePartitionStrategy(
+    ASTPtr partition_by_,
+    const Block & sample_block_,
+    ContextPtr context_,
+    const std::string & file_format_,
+    bool hive_partition_strategy_write_partition_columns_into_files_)
+    : PartitionStrategy(partition_by_, sample_block_, context_),
+    file_format(file_format_),
+    hive_partition_strategy_write_partition_columns_into_files(hive_partition_strategy_write_partition_columns_into_files_),
+    partition_expression_required_columns(extractPartitionRequiredColumns(partition_by, sample_block, context)),
+    partition_expression_required_columns_set(partition_expression_required_columns.begin(), partition_expression_required_columns.end())
+{
+    actions_with_column_name = buildExpressionHive(partition_by, partition_expression_required_columns, sample_block, context);
+    block_without_partition_columns = buildBlockWithoutPartitionColumns(sample_block, partition_expression_required_columns_set);
+}
+
+HiveStylePartitionStrategy::PartitionExpressionActionsAndColumnName HiveStylePartitionStrategy::getExpression()
+{
+    return actions_with_column_name;
+}
+
+std::string HiveStylePartitionStrategy::getPath(
+    const std::string & prefix,
+    const std::string & partition_key)
+{
+    std::string path;
+
+    if (!prefix.empty())
```

Comment:
I should probably add some logic here to prevent double slashes

Reply:
This is not a problem in my tests because AWS S3 sdk splits the key into several string using the slash delimiter, but we should be careful